File: gims_backend_api/src/main/java/com/kutub/InsuranceManagement/entity/marine/MarinePolicy.java
package com.kutub.InsuranceManagement.entity.marine;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.kutub.InsuranceManagement.entity.utility.Bank;
import jakarta.persistence.*;
import jakarta.validation.constraints.Positive;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.Date;
import java.util.List;

@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
@Table(name = "marine_insurance_details")
public class MarinePolicy {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @JsonProperty("policyId")
    @JsonAlias("id")
    private long id;

    @Column(unique = true)
    private String sysNumber;

    @Column(nullable = false)
    @Temporal(TemporalType.DATE)
    private Date date = new Date();

    @ManyToOne
    @JoinColumn(name = "bank_id", nullable = false)
    private Bank bank;

    @Column(nullable = false)
    private String policyholder;

    @Column(nullable = false)
    private String address;

    @Column(nullable = false)
    private String voyageFrom;

    @Column(nullable = false)
    private String voyageTo;

    @Column(nullable = false)
    private String via;

    @Column(nullable = false)
    private String stockItem;

    @Positive(message = "Sum insured in USD must be positive")
    @Column(nullable = false)
    private double sumInsuredUsd;

    @Column(nullable = false)
    private double usdRate;

    @Column(nullable = false)
    private double sumInsured;

    @Column(nullable = false)
    private String coverage;

    @JsonIgnore
    @OneToMany(mappedBy = "marineDetails", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<MarineBill> marineBills;

    public void convertSumInsuredUsd(double exchangeRate) {
        this.sumInsured = this.sumInsuredUsd * exchangeRate;
        this.usdRate = exchangeRate;
    }
}

File: gims_mobile_app/lib/models/marine/marine_policy.dart
import '../utility_models.dart';

class MarinePolicy {
  int? id;
  String? sysNumber;
  DateTime? date;
  Bank? bank;
  Branch? branch;
  String? policyholder;
  String? address;
  String? voyageFrom;
  String? voyageTo;
  String? via;
  String? stockItem;
  double? sumInsuredUsd;
  double? usdRate;
  double? sumInsured;
  String? coverage;
  bool isHovered = false;

  MarinePolicy({
    this.id,
    this.sysNumber,
    this.date,
    this.bank,
    this.branch,
    this.policyholder,
    this.address,
    this.voyageFrom,
    this.voyageTo,
    this.via,
    this.stockItem,
    this.sumInsuredUsd,
    this.usdRate,
    this.sumInsured,
    this.coverage,
  });

  factory MarinePolicy.fromJson(Map<String, dynamic> json) {
    double? parseDouble(dynamic value) {
      if (value == null) return null;
      if (value is num) return value.toDouble();
      if (value is String) return double.tryParse(value);
      return null;
    }

    return MarinePolicy(
      id: json['policy_id'] ?? json['id'],
      sysNumber: json['sys_number']?.toString(),
      date: json['date'] != null ? DateTime.tryParse(json['date'].toString()) : null,
      bank: json['bank'] != null ? Bank.fromJson(json['bank']) : null,
      branch: json['branch'] != null ? Branch.fromJson(json['branch']) : null,
      policyholder: json['policyholder']?.toString() ?? 'N/A',
      address: json['address']?.toString() ?? '',
      voyageFrom: (json['voyage_from'] ?? json['voyageFrom'])?.toString() ?? '',
      voyageTo: (json['voyage_to'] ?? json['voyageTo'])?.toString() ?? '',
      via: json['via']?.toString() ?? '',
      stockItem: (json['stock_item'] ?? json['stockItem'])?.toString() ?? '',
      sumInsuredUsd: parseDouble(json['sum_insured_usd'] ?? json['sumInsuredUsd']),
      usdRate: parseDouble(json['usd_rate'] ?? json['usdRate']),
      sumInsured: parseDouble(json['sum_insured'] ?? json['sumInsured']),
      coverage: json['coverage']?.toString() ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    if (id != null) data['id'] = id;
    if (sysNumber != null) data['sysNumber'] = sysNumber;
    if (date != null) data['date'] = date?.toIso8601String();
    if (bank != null) data['bank'] = {'id': bank!.id};
    if (branch != null) data['branch'] = {'id': branch!.id};
    data['policyholder'] = policyholder;
    data['address'] = address;
    data['voyageFrom'] = voyageFrom;
    data['voyageTo'] = voyageTo;
    data['via'] = via;
    data['stockItem'] = stockItem;
    data['sumInsuredUsd'] = sumInsuredUsd;
    data['usdRate'] = usdRate;
    data['sumInsured'] = sumInsured;
    data['coverage'] = coverage;
    return data;
  }
}
